invoices/resolution_result.go

```diff
@@ -0,0 +1,217 @@
+	// htlc does not match its set total.
+	ResultHtlcSetTotalMismatch
+
+	// ResultHtlcSetTotalTooLow is returned when a mpp set total is too low
+	// for an invoice.
+	ResultHtlcSetTotalTooLow
+
+	// ResultHtlcSetOverpayment is returned when a mpp set is overpaid.
+	ResultHtlcSetOverpayment
+
+	// ResultInvoiceNotFound is returned when an attempt is made to pay an
+	// invoice that is unknown to us.
+	ResultInvoiceNotFound
+
+	// ResultKeySendError is returned when we receive invalid keysend
+	// parameters.
+	ResultKeySendError
+)
+
+// ResultString returns a string representation of the result.
+//
+// Note: it is part of the ResolutionResult interface.
+func (f FailureResolutionResult) ResultString() string {
+	switch f {
+	case resultInvalidFailure:
+		return "invalid failure result"
+
+	case ResultReplayToCanceled:
+		return "replayed htlc to canceled invoice"
+
+	case ResultInvoiceAlreadyCanceled:
+		return "invoice already canceled"
+
+	case ResultAmountTooLow:
+		return "amount too low"
+
+	case ResultExpiryTooSoon:
+		return "expiry too soon"
+
+	case ResultCanceled:
+		return "cancelled"
+
+	case ResultInvoiceNotOpen:
+		return "invoice no longer open"
+
+	case ResultMppTimeout:
+		return "mpp timeout"
+
+	case ResultAddressMismatch:
+		return "payment address mismatch"
+
+	case ResultHtlcSetTotalMismatch:
+		return "htlc total amt doesn't match set total"
+
+	case ResultHtlcSetTotalTooLow:
+		return "set total too low for invoice"
+
+	case ResultHtlcSetOverpayment:
+		return "mpp is overpaying set total"
+
+	case ResultInvoiceNotFound:
+		return "invoice not found"
+
+	case ResultKeySendError:
+		return "invalid keysend parameters"
+
+	default:
+		return "unknown failure resolution result"
+	}
+}
+
+// SettleResolutionResult provides metadata which about an invoice update
+// that was failed by the registry. It can e used to take custom actions on
+// resolution of the htlc, only results which are actionable by the link are
+// exported.
+type SettleResolutionResult uint
+
+// ResultString returns a string representation of the result.
+//
+// Note: it is part of the ResolutionResult interface.
+func (s SettleResolutionResult) ResultString() string {
+	switch s {
+	case resultInvalidSettle:
+		return "invalid settle result"
+
+	case ResultSettled:
+		return "settled"
+
+	case ResultReplayToSettled:
+		return "replayed htlc to settled invoice"
+
+	case resultDuplicateToSettled:
+		return "accepting duplicate payment to settled invoice"
+
+	default:
+		return "unknown settle resolution result"
+	}
+}
+
+const (
+	resultInvalidSettle SettleResolutionResult = iota
+
+	// ResultSettled is returned when we settle an invoice.
+	ResultSettled
+
+	// ResultReplayToSettled is returned when we replay a settled invoice.
+	ResultReplayToSettled
+
+	// resultDuplicateToSettled is returned when we settle an invoice which
+	// has already been settled at least once.
+	resultDuplicateToSettled
```

Comment:
a little strange that the visibility of these variables is mixed, perhaps keep this public?